include/openspace/util/timemanager.h

```diff
@@ -32,41 +32,107 @@
 #include <utility>
 #include <vector>
 
+#include <deque>
+#include <functional>
+#include <openspace/util/timeline.h>
+#include <openspace/util/time.h>
+#include <openspace/util/syncdata.h>
+#include <openspace/properties/propertyowner.h>
+#include <openspace/properties/scalar/floatproperty.h>
+
 namespace openspace {
 
-class TimeManager {
+struct TimeKeyframeData {
+    Time time;
+    double delta;
+    bool pause = false;
+    bool jump = false;
+};
+
+class TimeManager : public properties::PropertyOwner {
 public:
+    TimeManager();
+
     using CallbackHandle = int;
     using TimeChangeCallback = std::function<void()>;
 
-    Time& time();
+    const Time& time() const;
+    const Time& integrateFromTime() const;
+    const Timeline<TimeKeyframeData>& timeline();
+
     std::vector<Syncable*> getSyncables();
     void preSynchronization(double dt);
-    void addKeyframe(double timestamp, Time keyframeTime);
-    void removeKeyframesBefore(double timestamp);
-    void removeKeyframesAfter(double timestamp);
-    void clearKeyframes();
+
+    TimeKeyframeData interpolate(double applicationTime);
+
     void setTimeNextFrame(Time t);
+    void setDeltaTime(double deltaTime);
+    void setPause(bool pause);
+
+    double targetDeltaTime() const;
+    double deltaTime() const;
+    bool isPaused() const;
+
+    float defaultTimeInterpolationDuration() const;
+    float defaultDeltaTimeInterpolationDuration() const;
+    float defaultPauseInterpolationDuration() const;
+    float defaultUnpauseInterpolationDuration() const;
+
+    void interpolateTime(double targetTime, double durationSeconds);
+    void interpolateDeltaTime(double targetDeltaTime, double durationSeconds);
+    void interpolatePause(bool pause, double durationSeconds);
+
+    void addKeyframe(double timestamp, TimeKeyframeData kf);
+    void removeKeyframesBefore(double timestamp, bool inclusive = false);
+    void removeKeyframesAfter(double timestamp, bool inclusive = false);
+
     size_t nKeyframes() const;
+    void clearKeyframes();
 
     CallbackHandle addTimeChangeCallback(TimeChangeCallback cb);
     CallbackHandle addDeltaTimeChangeCallback(TimeChangeCallback cb);
+    CallbackHandle addTimeJumpCallback(TimeChangeCallback cb);
+    CallbackHandle addTimelineChangeCallback(TimeChangeCallback cb);
+
     void removeTimeChangeCallback(CallbackHandle handle);
     void removeDeltaTimeChangeCallback(CallbackHandle handle);
-
+    void removeTimeJumpCallback(CallbackHandle handle);
+    void removeTimelineChangeCallback(CallbackHandle handle);
 private:
+    void progressTime(double dt);
+    void applyKeyframeData(const TimeKeyframeData& keyframe);
+    TimeKeyframeData interpolate(
+        const Keyframe<TimeKeyframeData>& past,
```

Comment:
For style:  This line can be in the same line as the function definition;  the `time` can be moved in with the `future` line, too